### PR TITLE
Added dao for users, apps, event_types, and events 

### PR DESCRIPTION
**Pull request by noobbyte**
With @young-k

### CODE REVIEW

File: src/app/events/utils/db_utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from . import *
```
Comment (Jma353) [inline src/app/events/utils/db_utils.py]: This is repeated from podcasts - we can move this to shared pip module `appdev.py` this way we don't need to recreate it.  We can integrate this updated module in a separated PR, but add a `# TODO` indicating we should do this

File: src/app/events/models/user.py [hunk 2]
```diff
@@ -25,10 +25,10 @@ class User(Base):
...
-    self.password = bcrypt.hashpw(kwargs.get('password'), bcrypt.gensalt())
+    self.password_digest = bcrypt.hashpw(kwargs.get('password').encode('utf8'),
+                                         bcrypt.gensalt())
     self.first_name = kwargs.get('first_name')
     self.last_name = kwargs.get('last_name')
 
   def verify_password(self, password):
-    pwhash = bcrypt.hashpw(password, self.password)
-    return self.password == pwhash
+    return bcrypt.checkpw(password.encode('utf8'), self.password_digest.encode('utf8'))
```
Comment (Jma353) [inline src/app/events/models/user.py]: Is this correct - did we check this?  Just making sure.  Seems in line with everything [this](http://dustwell.com/how-to-handle-passwords-bcrypt.html) guide says.  Might want to specify log rounds to be  high on the gen salt, as we will have very few users using this, but would like to maximize security.

File: src/app/events/models/event_type.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 class EventType(Base):
```
Comment (Jma353) [inline src/app/events/models/event_type.py]: We should be relating users to this somehow, so we can track who made what `event_types`.  Given that this `PR` involves users, we should specify a foreign key on `users` (`user_id`) as well, and remove `creator`, as that was temporary.

File: src/app/events/dao/users_dao.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  return optional_user.verify_password(password)
+
+def create_user(email, password, first_name='', last_name=''):
+  optional_user = get_user_by_email(email)
+  
+  if optional_user is not None:
+    return False, optional_user
+
+  # user does not exist
+  user = User(email=email, password=password,
```
Comment (Jma353) [inline src/app/events/dao/users_dao.py]: Have these `kwargs` go line by line.  

```python
user = User(
  email=email,
  password=password,
   # etc.
)
```

File: src/app/events/dao/users_dao.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+              first_name=first_name, last_name=last_name)
+  db_utils.commit_model(user)
+  return True, user
+
+def get_apps(user_id):
```
Comment (Jma353) [inline src/app/events/dao/users_dao.py]: `get_users_apps`

File: src/app/events/dao/applications_dao.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+from . import *
```
Comment (Jma353) [inline src/app/events/dao/applications_dao.py]: Add a function `is_owned_by_user` to see if a user is one of the owners of an `app`.

File: src/app/events/dao/event_types_dao.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    return False, event_type
+  
+  verify_fields_info(fields_info)
+
+  app = applications_dao.get_app_by_id(app_id)
+
+  if app is None:
+    raise Exception('App does not exist.')
+  
+  event_type = EventType(name=name, creator=creator,
```
Comment (Jma353) [inline src/app/events/dao/event_types_dao.py]: `kwargs` line-by-line

File: src/app/events/dao/events_dao.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+def create_event(event_type_id, payload):
+  event_type = event_types_dao.get_event_type_by_id(event_type_id)
+  
+  if event_type is None:
+    raise Exception('Event type does not exist.')
+
+  event_types_dao.verify_fields(event_type_id, payload)
+  event = Event(event_type_id=event_type_id, payload=payload)
+  db_utils.commit_model(event)
+  return True, event
```
Comment (Jma353) [inline src/app/events/dao/events_dao.py]: If we return `True, event` here, we should return `False, ...` somewhere else on some exception handling:

```python
try:
  db_utils.commit_model(event)
  return True, event
except Exception as e:
  return False, e
```

File: src/app/events/dao/event_types_dao.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  db_utils.commit_model(event_type)
+  return True, event_type
+
+def get_event_type_by_name(app_id, name):
+  return EventType.query.filter(EventType.application_id == app_id, EventType.name == name).first()
+
+def get_event_type_by_id(event_type_id):
+  return EventType.query.filter(EventType.id == event_type_id).first()
+
+def verify_fields_info(fields_info):
```
Comment (Jma353) [inline src/app/events/dao/event_types_dao.py]: I'm not really a fan of this branching if statement - this validation should be more of state-machine, with checks that happen regardless of whether fields exist in metadata, and pass along result of the next check if the current check is `True`, and `False if the current check is `False`.  That being said, I think we still might refine how this metadata / validation process occurs in the future, so feel free to keep this but mark it with a `# TODO`, what you have written rn will service our current needs fine.

File: src/app/events/dao/event_types_dao.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+  if event_type is None:
+    raise Exception('Event type does not exist.')
+  
+  fields_info = event_type.fields_info
+
+  for field in fields_info:
+    metadata = fields_info[field]
+
+    if field not in payload:
```
Comment (Jma353) [inline src/app/events/dao/event_types_dao.py]: Same this as above - state-machine, but keep this and add `# TODO` for now.

**Comment by Jma353** [comment]
Excellent work!  Left some comments, not really crazy amount of change, just some things I noticed.  Please address those comments in a follow-up commit and feel free to merge after that .

**Comment by young-k** [comment]
15fa30169096ea930d387947cade05522e1c4006

**Pull request merged**